compiler/optimizations/inferConstRefs.cpp

```diff
@@ -350,59 +411,281 @@ static bool inferConstRef(Symbol* sym) {
+    return wasRefToConst;
+  }
+
+  bool isFirstCall = false;
+  if (info && info->alreadyCalled == false) {
+    isFirstCall = true;
+    info->alreadyCalled = true;
+  }
+
+  bool isRefToConst = true;
+
+  if (isArgSymbol(sym)) {
+    // Check each call and set isRefToConst to false if any actual is not a ref
+    // to a const.
+    FnSymbol* fn = toFnSymbol(sym->defPoint->parentSymbol);
+    if (fn->hasFlag(FLAG_VIRTUAL)) {
+      // Not sure how to best handle virtual calls, so simply set false for now
+      isRefToConst = false;
     } else {
-      INT_ASSERT(isVarSymbol(sym));
-      sym->qual = QUAL_CONST_REF;
+      if (isFirstCall && info) {
+        INT_ASSERT(info->fnUses == NULL);
+        info->fnUses = fn->firstSymExpr();
+      } else if (info == NULL) {
+        isRefToConst = false;
+      }
+      // Need this part to be re-entrant in case of recursive functions
+      while (info && info->fnUses != NULL && isRefToConst) {
+        SymExpr* se = info->fnUses;
+        info->fnUses = se ? se->symbolSymExprsNext : NULL;
+
+        CallExpr* call = toCallExpr(se->parentExpr);
+        INT_ASSERT(call && call->isResolved());
+
+        Symbol* actual = toSymExpr(formal_to_actual(call, sym))->symbol();
+
+        if (actual->isRef()) {
+          // I don't think we technically need to skip if the actual is the
+          // same symbol as the formal, but it makes things simpler.
+          if (actual != sym && !inferRefToConst(actual)) {
+            isRefToConst = false;
+          }
+        } else {
+          // Passing a non-ref actual to a reference formal is currently
+          // considered to be the same as an addr-of
+          if (actual->qualType().getQual() != QUAL_CONST_VAL) {
+            isRefToConst = false;
+          }
+        }
+      }
+    }
+  }
+
+  while (info && info->hasMore() && isRefToConst) {
+    SymExpr* use = info->next();
+
+    CallExpr* call = toCallExpr(use->parentExpr);
+    if (call == NULL) continue;
+
+    if (isMoveOrAssign(call)) {
+      if (use == call->get(1)) {
+        if (SymExpr* se = toSymExpr(call->get(2))) {
+          if (se->isRef() && !inferRefToConst(se->symbol())) {
+            isRefToConst = false;
+          }
+        }
+        else {
+          CallExpr* RHS = toCallExpr(call->get(2));
+          INT_ASSERT(RHS);
+          if (RHS->isPrimitive(PRIM_ADDR_OF) ||
+              RHS->isPrimitive(PRIM_SET_REFERENCE)) {
+            SymExpr* src = toSymExpr(RHS->get(1));
+            if (src->isRef()) {
+              if (!inferRefToConst(src->symbol())) {
+                isRefToConst = false;
+              }
+            } else {
+              if (src->symbol()->qualType().getQual() != QUAL_CONST_VAL) {
+                isRefToConst = false;
+              }
+            }
+          } else {
+            isRefToConst = false;
+          }
+        }
+      }
+    }
+    else if (call->isResolved()) {
+      isRefToConst = true;
```

Comment:
The case above is a formal looking at its corresponding actuals, because a formal can't be ref-to-const unless all of its actuals are const-to-ref.

When 'sym' is an actual, in the case you've commented on here, we don't care whether or not the formal is ref-to-const.